plutus-tx/src/PlutusTx/Extensions/Secrets.hs

```diff
@@ -0,0 +1,133 @@
+ - imports `PlutusTx.Extensions.Secrets` does not use some unholy `unsafePerformIO` to
+ - break the Haskell type system the code is guaranteed not to depend on the actual
+ - value of a secret without:
+ -   1. Using a safe function like `escape_sha2_256` or
+ -   2. Using an unsafe function like `unsafe_escape_secret` (which has a compiler warning
+ -      attached to it).
+ -
+ - This intuition can be made formally precise (see Algehed and Bernardy 2019).
+ -
+ - As noted above, `unsafe_escape_secret` breaks the abstraction barrier and we need to export
+ - it for two reasons:
+ -   1. To make the library backwards-compatible. If you have some new hash function you should
+ -      be able to define `escape_my_magic_hash` without having to implement it in this module.
+ -   2. For tests it is sometimes necessary to use `unsafe_escape_secret` to inspect the value
+ -      of a secret.
+ -
+ - See `Plutus.Contracts.SealedBidAuction` or `Plutus.Contract.GameStateMachineWithSecretTypes`
+ - for examples.
+ -}
+
+-- | Secret argments are provided in the endpoint argument types.
+--
+-- This type guarantees that a `SecretArgument a` that is seen by
+-- the endpoint code is a `Secret a` in a way that can not be
+-- bypassed by safe code.
+data SecretArgument a = UserSide a
+                      | EndpointSide (Secret a)
+                      deriving Show
+
+{- Note [Secret arguments]
+ - When we write endpoint code we would like to specify the argument type
+ - for an endpoint with code that looks something like this:
+ - ```
+ - data MyEndpointArgs = MyEndpointArgs { publicArg :: Int , secretArg :: Secret Int }
+ -
+ - type UserSchema = Endpoint "myendpoint" MyEndpointArgs
+ -                   .\/ ...
+ - ```
+ - Which would mean that any endpoint code for `myendpoint` would be guaranteed not to
+ - leak the `secretArg` in plaintext to the blockchain (or elsewhere for that matter).
+ -
+ - However, the type `MyEndpointArgs` needs to be an instance of the class `ToJSON`
+ - for this to work out with the `endpoint` function. Consequently, we would need to
+ - provide a function `toJSON :: Secret a -> Value` that would have to break abstraction
+ - for the `Secret` type (see [Secret abstraction]) which would be bad.
+ -
+ - The `SecretArgument a` type fixes this by having two constructors `UserSide :: a -> SecretArgument a`
+ - and `EndpointSide :: Secret a -> SecretArgument a` that ensures that as secrets are submitted
+ - to endpoint code by the user they are "public", while when the secret reaches the endpoint code it
+ - has been obscured by a `Secret` wrapper that gives us the guarantee that the secret doesn't leak.
+ -
+ - See `Plutus.Contracts.SealedBidAuction` or `Plutus.Contract.GameStateMachineWithSecretTypes` for examples.
+ -}
+
+instance ToJSON a => ToJSON (SecretArgument a) where
+  toJSON (UserSide a)     = toJSON a
+  toJSON (EndpointSide _) = Aeson.String "EndpointSide *****"
```

Comment:
So this deliberately doesn't roundtrip? Seems a little questionable...

Reply:
This is unfortunately the only way to make this work out with the way that `endpoint` is implemented. One possible fix for this would be to rewrite `endpoint` to teach it about secrets. I don't know if this is better or worse as it would require a separate type class for applying the "hiding" transformation.

Reply:
For any outside observer's reference: this is explained a bit better in the note [Secret arguments].

Reply:
I'm open to fixing this by working around it in `endpoint` - do you have any intuition for what the least obtrusive way of doing so would be @michaelpj?

Reply:
Not really. @j-mueller ?

Reply:
One possible approach to solving this is to do something like add a class `IsEndpointArgument a` with a single function `scrubSecrets :: IsEndpointArgument a => a -> a` that is a homomorphism w.r.t. all constructors of types other than `SecretArgument a` where it is defined as:
```
scrubSecret (UserSide a) = EndpointSide (mkSecret a)
scrubSecret (EndpointSide s) = EndpointSide s
```
which would result in having to change the type of `endpoint` to:
```
endpoint
  :: forall l a w s e b.
     ( HasEndpoint l a s
     , AsContractError e
     , FromJSON a
     , IsEndpointArgument a
     )
  => (a -> Contract w s e b) -> Promise w s e b
```
Of course, one could also fold `FromJSON a` into this to have only a constraint `IsEndpointArgument a` on `a`.

Regardless, I'm not sure it's nicer to impose this solution as opposed to this deliberate hack that gets around having to change the whole `endpoint` interface (and all examples, tutorials, etc).

Reply:
Well if it's a choice between changing the endpoint interface to support this, and having a dodgy JSON instance, then I would prefer the instance ;)

Reply:
That means it makes sense to keep it this way until the endpoint interface is next revised and use that as an opportunity to restructure this code too? The type:
```
endpoint
  :: forall l a w s e b.
     ( HasEndpoint l a s
     , AsContractError e
     , IsEndpointArgument a
     )
  => (a -> Contract w s e b) -> Promise w s e b
```
with a suitable definition of `IsEndpointArgument` and a TH function for deriving `IsEndpointArgument` instances is perhaps more general and extensible than the current interface - but it is of course a breaking change.